lib/Onyx.js

```diff
@@ -412,48 +402,6 @@ function connect(mapping) {
     return connectionID;
 }
 
-/**
- * Remove cache items that are no longer connected through Onyx
- * @param {string} key
- */
-function cleanCache(key) {
-    // Don't remove default keys from cache, they don't take much memory and are accessed frequently
-    if (_.has(defaultKeyStates, key)) {
-        return;
-    }
-
-    const hasRemainingConnections = _.some(callbackToStateMapping, {key});
-
-    // When the key is still used in other places don't remove it from cache
-    if (hasRemainingConnections) {
```

Comment:
Just to confirm after this change - if we theoretically have something like 200 keys simultaneously subscribed and disconnected just one that we'd drop 50 of the least recently accessed whether a subscriber is accessing them or not?

The previous logic only drops keys that are not being subscribed to so I was surprised to see that removed. Just curious to hear thought about why we shouldn't need logic like that anymore since it it being removed.

Reply:
> Just to confirm after this change - if we theoretically have something like 200 keys simultaneously subscribed and disconnected just one that we'd drop 50 of the least recently accessed whether a subscriber is accessing them or not?

Yes this is correct, but it would be an extremely rare case for this to happen
IRL we do have 200+ connections but they involve about 50-60 unique keys

> The previous logic only drops keys that are not being subscribed to so I was surprised to see that removed. Just curious to hear thought about why we shouldn't need logic like that anymore since it it being removed.

Good point. I forgot to bring that up

In short if something is "so old" yet still connected after we've cycled 100-150 keys and it didn't re-appear at least ones to revive itself - it's not something that will benefit from the "cache while connected" strategy - no other connection/component wants to use it anyway

---

The limit of 150 keys is actually too big. I've browsed storage on Android and see that for my account with 10 chats I have 58 unique keys in storage so I'll never reach the LRU limit of 150
Of these 58 keys only 10 can grow indefinitely - `reportActions_...` (I guess that's why they are the "eviction keys") the rest are of finite and small size - precisely 34.1kB (24.1kB if we go and remove the report objects)

With this information I think we should consider 3 items for the future:
1. Don't drop non eviction keys from cache - these are the frequently used keys and often all of them are in memory. 
2. Load all non-eviction keys in one multiGet during init
3. Apply the LRU strategy only for the eviction keys

#### Here's what I'm planning for the current PR:

Reduce the LRU limit to 50-60 keys - this will still let cache fill with all the reusable keys during init (48 keys) 
Cache will start to free up after you switch between some chats

Reply:
Good thoughts thanks for the explanation. If it helps at all, I have 310 keys in `localStorage` for my main account that is maybe more of a typical business usage.

It seems reasonable to go with something more aggressive that has an arbitrary limit for now (which seems better than what we had before).

It does feel like the most important case is switching to a chat we've already accessed in a session and will need to retrieve the `reportActions_*` from storage instead of cached (grabbing and parsing the big object will take time). So this should help with that.

I do agree it seems less value to drop anything finite and may be better to front load the cost of getting finite keys to a `multiGet` on init and then just never drop those keys.

This is maybe thinking ahead a bit. But if we do go that direction we can potentially grab values synchronously for `withOnyx` connected components that are only interested in cached values - which might be a huge improvement for `withOnyx`. But that's only possible if we can guarantee which keys will be cached and which are subject to eviction.

Reply:
> I do agree it seems less value to drop anything finite and may be better to front load the cost of getting finite keys to a multiGet on init and then just never drop those keys.

I'll open follow up ticket to discuss the points above

> This is maybe thinking ahead a bit. But if we do go that direction we can potentially grab values synchronously for withOnyx connected components that are only interested in cached values - which might be a huge improvement for withOnyx. But that's only possible if we can guarantee which keys will be cached and which are subject to eviction.

Yes, that sounds about right 
We can use the deferred task, and release it after we've prefilled cache with the non-eviction keys. This will happens faster than letting components asks for these keys one by one during init

Now `withOnyx` can pick values synchronously from cache
Hmm, but actually only the connections are waiting `init` to resolve
Components would mount right away and initiate a connection during the `didMount` phase - component construction and initial state won't wait for Onyx.init
There are ways around this though - like rendering the WrappedComponent until Onyx finishes init and then passing the state in one update

BTW it might be better to connect to Onyx in the constructor instead of during `componentDidMount`. 
In `didMount` it means that at least one commit happened and only then we start the connection